### ISSUE TITLE
[cmake] Secondary dependencies HDF5 not propagated when building static lib

### ISSUE THREAD

**Issue by simogasp** [alembic/alembic#126]
Hi @lamiller0 

when one tries to build alembic as a static lib (`-DALEMBIC_SHARED_LIBS:BOOL=OFF`) and with HDF5 support (`-DUSE_HDF5:BOOL=ON`), then the linking with alembic **as third party** fails because hdf5 is not added to the list of libraries (i.e. undefined symbols for x86_64 kind of errors). This occurs because `TARGET_LINK_LIBRARIES` is not set correctly, it should read
```cmake
TARGET_LINK_LIBRARIES(Alembic
    PUBLIC
    ${ALEMBIC_ILMBASE_LIBS}
    PRIVATE
    ${HDF5_LIBRARIES}
  )
```
instead of having just `${ALEMBIC_ILMBASE_LIBS}`.  It must be set to `PRIVATE` because HDF5 is not a public dependency, yet when building static it must be propagated.
This change allows cmake to correctly propagate the secondary dependencies in `AlembicTargets.cmake`. 
I noticed that few lines below you have some code like this
```cmake
# fixes undefined symbols for x86_64
IF (ALEMBIC_SHARED_LIBS AND DARWIN)
    TARGET_LINK_LIBRARIES(${CORE_LIBS})
ENDIF()
```
This is exactly a workaround to the fact that hdf5 is not listed among the above linked libraries (`${CORE_LIBS}` has it though), it allows to build the library but not to set up correctly the  `AlembicTargets.cmake` for correctly linking as 3rd party. Removing this workaround and setting up TARGET_LINK_LIBRARIES as described above made everything work for me (both static and dynamic). 

I can do a PR with this fix and few more fixes concerning cmake (see other issue  #127)

Simone

**Comment by lamiller0** [comment]
I look forward to seeing your PR.

**Comment by simogasp** [comment]
@lamiller0 
sorry, I just noticed that the minimum required version for cmake is currently 2.8.11 and `PUBLIC`/`PRIVATE` inside `TARGET_LINK_LIBRARIES` has been introduced in 2.8.12. 
There are two options:
1. use `LINK_PUBLIC` and `LINK_PRIVATE` instead (https://cmake.org/cmake/help/v2.8.11/cmake.html#command:target_link_libraries)
2. update the minimum required version to 2.8.12 (the keywords `LINK_PUBLIC` and `LINK_PRIVATE` are considered legacy by newer versions of cmake)

I'd be more in favour of 2. (cmake 2.8.12 was already shipped with eg ubuntu 14.04), but I do get that there may be other people who might not be able to use higher versions of cmake for some reasons. What do you think?

Also, diving a little bit more inside the cmake files i noticed that we can get completely rid of`CORE_LIBS`, and replace it with the target Alembic wherever it is used (which automagically will include all the necessary secondary dependencies). I tested it with multiple configurations for the different compiling options and it works both on osx and linux (sorry i cannot test on windoze, and with houdini, arnold and prman but it should not be breaking anything)

Just let me know what you think and I will update and clean the PR.

S.

**Comment by lamiller0** [comment]
I'd like to stay with the minimum CMake 2.8.11 for this PR.

Reaching out to the community about what a reasonable new minimum is, and modernizing the project accordingly should be a separate issue and PR.

Feel free to include the `CORE_LIBS` to `Alembic` target changes, I can do some simple tests on Windows.

**Comment by simogasp** [comment]
@lamiller0 
Ok, I clearly understand your point.

I rebased the PR replacing `PUBLIC` / `PRIVATE` with `LINK_PUBLIC` and `LINK_PRIVATE`.
I also removed `CORE_LIBS` and replaced with `Alembic` target. To be on the safe side, I kept the dependency thread and libm. If I'm not wrong those are not direct dependencies of Alembic, but they are the second dependencies of Ilmbase libs. In theory it should be up to the FindIlmbase to deal with that so that the second dependencies are included transparently and correctly propagated. That could be taken care in a proper modernization refactoring of the project, as you said. Since i cannot test it on all platforms i'd keep them, they will do no harm anyways.

Let me know how your tests with windows go :-)

S.

PS
if you want to check, I also tested different compilers with the different options on travis
https://travis-ci.org/alicevision/alembic/builds/231109441
if you want, i can make a different PR for that updated travis script, it tests gcc 4.8 and 5, clang 3.5 and 3.8, with and without hdf5, building static and dynamic (it takes ~30min to complete everything). I'd be happy to do another PR for that if you considered it interesting.
